src/closure.rs

```diff
@@ -0,0 +1,136 @@
+// Copyright 2013-2017, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+// TODO: support marshaller.
+
+use std::mem;
+use std::ptr;
+use std::slice;
+
+use libc::{c_uint, c_void};
+
+use gobject_ffi;
+use translate::{ToGlibPtr, ToGlibPtrMut, Uninitialized, from_glib_none};
+use types::Type;
+use Value;
+use ToValue;
+
+glib_wrapper! {
+    pub struct Closure(Shared<gobject_ffi::GClosure>);
+
+    match fn {
+        ref => |ptr| {
+            gobject_ffi::g_closure_ref(ptr);
+            gobject_ffi::g_closure_sink(ptr);
+        },
+        unref => |ptr| gobject_ffi::g_closure_unref(ptr),
+    }
+}
+
+impl Closure {
+    pub fn new<F: Fn(&[Value]) -> Option<Value> + 'static>(callback: F) -> Self {
```

Comment:
One problem I forgot about is that GClosure is `Sync+Send`. You have no guarantees from which thread your callback is called, you have no guarantees that it isn't called simultaneously from multiple threads at once. As such this should probably be `F: Fn(&[Value]) -> Option<Value> + Send + Sync + 'static`